lib/rich_text.rb

```diff
@@ -87,17 +87,60 @@ def sanitize(text)
     end
 
     def linkify(text, mode = :urls)
-      link_attr = 'rel="nofollow noopener noreferrer" dir="auto"'
-      Rinku.auto_link(ERB::Util.html_escape(text), mode, link_attr) do |url|
-        url = shorten_host(url, Settings.linkify_hosts, Settings.linkify_hosts_replacement)
-        shorten_host(url, Settings.linkify_wiki_hosts, Settings.linkify_wiki_hosts_replacement) do |path|
-          path.sub(Regexp.new(Settings.linkify_wiki_optional_path_prefix || ""), "")
-        end
-      end.html_safe
+      ERB::Util.html_escape(text)
+               .then { |html| expand_link_shorthands(html) }
+               .then { |html| expand_host_shorthands(html) }
+               .then { |html| auto_link(html, mode) }
+               .html_safe
     end
 
     private
 
+    def expand_link_shorthands(text)
+      Array.wrap(Settings.linkify&.detection_rules)
+           .select { |rule| rule.path_template && rule.patterns.is_a?(Array) }
+           .flat_map { |rule| linkify_detection_rule_to_gsub(rule) }
```

Comment:
Is `flat_map` correct here? We want a list of pattern and replacement pairs and `linkify_detection_rule_to_gsub` returns a pattern/replacement pair so running it on each rule will give us what we want and flattening will destroy that?